src/test/java/com/google/jenkins/plugins/computeengine/integration/ComputeClientV2IT.java

```diff
@@ -0,0 +1,146 @@
+import com.google.jenkins.plugins.computeengine.client.ComputeClientV2;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import lombok.extern.java.Log;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+@Log
+public class ComputeClientV2IT {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    private static final Map<String, String> googleLabels = getLabel(ComputeClientV2IT.class);
+    private static final Map<String, String> nonJenkinsLabels = Map.of("non-jenkins-label", "non-jenkins-value");
+    private static final Map<String, String> newLabels = Map.of("new-key", "new-value");
+
+    @Before
+    public void setUp() throws Exception {
+        initCredentials(j);
+        initCloud(j);
+    }
+
+    @After
+    public void tearDown() throws IOException {
+        teardownResources(getCloud(j).getClient(), googleLabels, log);
+        teardownResources(getCloud(j).getClient(), nonJenkinsLabels, log);
+        // merge googleLabels and newLabels for teardown purposes only
+        Map<String, String> allLabels = new HashMap<>();
+        allLabels.putAll(googleLabels);
+        allLabels.putAll(newLabels);
+        teardownResources(getCloud(j).getClient(), allLabels, log);
+    }
+
+    public ComputeEngineCloud getCloud(JenkinsRule j) {
+        return (ComputeEngineCloud) j.jenkins.clouds.getByName("gce-integration");
+    }
+
+    // create dummy instances
+    public String createOneInstance(ComputeClientV2 clientV2, Map<String, String> googleLabels) throws IOException {
+        var instanceConfig = instanceConfigurationBuilder()
+                .numExecutorsStr(NUM_EXECUTORS)
+                .labels(LABEL)
+                .oneShot(false)
+                .createSnapshot(false)
+                .template(NULL_TEMPLATE)
+                .zone(ZONE)
+                .googleLabels(googleLabels)
+                .build();
+        Instance instance = instanceConfig.instance();
+        clientV2.getCompute().instances().insert(PROJECT_ID, ZONE, instance).execute();
+        return instance.getName();
+    }
+
+    @Test
+    public void testRetrieveInstancesByLabelAndStatus() throws Exception {
+        ComputeClientV2 clientV2 = ComputeClientV2.createFromComputeEngineCloud(getCloud(j));
+        String instance1 = createOneInstance(clientV2, googleLabels);
+        String instance2 = createOneInstance(clientV2, nonJenkinsLabels);
+        await().atMost(2, TimeUnit.MINUTES).until(() -> {
+            var i1 = getInstance(clientV2, instance1);
+            var i2 = getInstance(clientV2, instance2);
+            return i1 != null
+                    && i2 != null
+                    && i1.getStatus().equals("RUNNING")
+                    && i2.getStatus().equals("RUNNING");
+        });
```

Comment:
Could be written with a type-safe matcher

```suggestion
        await().atMost(2, TimeUnit.MINUTES).until(() -> List.of(getInstance(clientV2, instance1), getInstance(clientV2, instance2)), Every.everyItem(instanceIsRunning()));
```

and 

```
    private Matcher<Instance> instanceIsRunning() {
        return new RunningStatusMatcher();
    }

    private static class RunningStatusMatcher extends TypeSafeMatcher<Instance> {
        @Override
        protected boolean matchesSafely(Instance instance) {
            return instance != null && instance.getStatus().equals("RUNNING");
        }

        @Override
        public void describeTo(Description description) {
            description.appendText("instance status is RUNNING");
        }
    }
```

Reply:
wow.. thank you